File: spring_java/src/main/java/dev/pollito/spring_java/sakila/auth/domain/service/AuthUseCasesImpl.java
package dev.pollito.spring_java.sakila.auth.domain.service;

import static java.util.Objects.requireNonNull;
import static org.springframework.security.authentication.UsernamePasswordAuthenticationToken.unauthenticated;
import static org.springframework.security.core.context.SecurityContextHolder.getContext;

import dev.pollito.spring_java.config.security.jwt.JwtService;
import dev.pollito.spring_java.config.security.userdetails.SakilaUserDetails;
import dev.pollito.spring_java.sakila.auth.domain.port.in.AuthUseCases;
import lombok.RequiredArgsConstructor;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class AuthUseCasesImpl implements AuthUseCases {

  private final AuthenticationManager authenticationManager;
  private final JwtService jwtService;

  @Override
  public String authenticate(String username, String password) {
    return jwtService.generateToken(
        (UserDetails)
            requireNonNull(
                authenticationManager
                    .authenticate(unauthenticated(username, password))
                    .getPrincipal()));
  }

  @Override
  public SakilaUserDetails getCurrentUser() {
    Authentication authentication = getContext().getAuthentication();
    if (authentication == null
        || !(authentication.getPrincipal() instanceof SakilaUserDetails details)) {
      throw new IllegalStateException("No authenticated user found");
    }
    return details;
  }
}

File: spring_kotlin/src/main/kotlin/dev/pollito/spring_kotlin/sakila/auth/domain/service/AuthUseCasesImpl.kt
package dev.pollito.spring_kotlin.sakila.auth.domain.service

import dev.pollito.spring_kotlin.config.security.jwt.JwtService
import dev.pollito.spring_kotlin.config.security.userdetails.SakilaUserDetails
import dev.pollito.spring_kotlin.sakila.auth.domain.port.`in`.AuthUseCases
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Service

@Service
class AuthUseCasesImpl(
    private val authenticationManager: AuthenticationManager,
    private val jwtService: JwtService,
) : AuthUseCases {

  override fun authenticate(username: String, password: String): String {
    val authentication =
        authenticationManager.authenticate(
            UsernamePasswordAuthenticationToken.unauthenticated(username, password)
        )
    val userDetails = authentication.principal as SakilaUserDetails
    return jwtService.generateToken(userDetails)
  }

  override fun getCurrentUser(): SakilaUserDetails {
    val authentication = SecurityContextHolder.getContext().authentication
    require(authentication != null && authentication.principal is SakilaUserDetails) {
      "No authenticated user found"
    }
    return authentication.principal as SakilaUserDetails
  }
}
